provider/equinix/environ.go

```diff
@@ -0,0 +1,665 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package equinix
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/juju/cloudconfig/cloudinit"
+	"github.com/juju/juju/cloudconfig/instancecfg"
+	"github.com/juju/juju/cloudconfig/providerinit"
+	"github.com/juju/juju/core/constraints"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/environs"
+	environscloudspec "github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/imagemetadata"
+	"github.com/juju/juju/environs/instances"
+	envinstance "github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/environs/simplestreams"
+	"github.com/juju/juju/provider/common"
+	"github.com/juju/juju/storage"
+	"github.com/juju/juju/tools"
+	"github.com/juju/loggo"
+	"github.com/juju/os/series"
+	"github.com/juju/retry"
+	"github.com/juju/schema"
+	"github.com/juju/utils/arch"
+	"github.com/juju/utils/set"
+	"github.com/juju/version"
+	"gopkg.in/juju/environschema.v1"
+
+	"github.com/packethost/packngo"
+)
+
+var logger = loggo.GetLogger("juju.provider.equnix")
+
+type environConfig struct {
+	config *config.Config
+	attrs  map[string]interface{}
+}
+
+type environ struct {
+	ecfgMutex    sync.Mutex
+	ecfg         *environConfig
+	name         string
+	cloud        environscloudspec.CloudSpec
+	equnixClient *packngo.Client
+	namespace    instance.Namespace
+}
+
+// func newEMEnviron() (*environ, error) {
+// 	e := new(environ)
+// 	namespace, err := instance.NewNamespace(e.ecfg.config.UUID())
+// 	if err != nil {
+// 		return nil, errors.Trace(err)
+// 	}
+
+// 	e.ecfg, err = providerInstance.newConfig(cfg)
+// 	if err != nil {
+// 		return nil, errors.Trace(err)
+// 	}
+
+// 	e.namespace = namespace
+// 	return e, nil
+// }
+
+var providerInstance environProvider
+
+func (e *environ) AdoptResources(ctx context.ProviderCallContext, controllerUUID string, fromVersion version.Number) error {
+	return nil
+}
+
+func (e *environ) Bootstrap(ctx environs.BootstrapContext, callCtx context.ProviderCallContext, args environs.BootstrapParams) (*environs.BootstrapResult, error) {
+	return common.Bootstrap(ctx, e, callCtx, args)
+}
+
+func (e *environ) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tags map[string]string) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	packetTags := []string{}
+
+	for k, v := range tags {
```

Comment:
There is no need to prepare the slice and then create a set out of it. You could simplify that as follows:

```suggestion
        deviceTags := set.NewStrings()
	for k, v := range tags {
	   deviceTags.Add(fmt.Sprintf("%s=%s", k, v))
	 }
```